### PR TITLE
ui: port syspolicy handler code to new app

### PR DESCRIPTION
**Pull request by kari-ts**
port over https://github.com/tailscale/tailscale-android/pull/199 from cmd/tailscale and legacy_android to libtailscale and android/

fix issue where we were creating MDMSettings every time we were retrieving policy settings 

Updates tailscale/corp#18202

### CODE REVIEW

File: android/src/main/java/com/tailscale/ipn/mdm/MDMSettingsDefinitions.kt [hunk 1]
```diff
@@ -4,14 +4,32 @@
...
+  }
   val flow: StateFlow<T> = MutableStateFlow<T>(defaultValue)
 
+  private fun registerKey(key: String) {
+    if (!keyExists(key)) {
+        keys.add(key)
+    }
+}
+
+fun keyExists(key: String): Boolean {
```
Comment (barnstar) [inline android/src/main/java/com/tailscale/ipn/mdm/MDMSettingsDefinitions.kt:29]: nit: the one line method here isn't really needed (unless it's called by something else I'm not seeing.)

**Review by barnstar — APPROVED** [review]
Looks good, but @agottardo should take a look as well.

File: android/src/main/java/com/tailscale/ipn/mdm/MDMSettingsDefinitions.kt [hunk 2]
```diff
@@ -4,14 +4,32 @@
...
 
 import android.os.Bundle
+import android.util.Log
 import com.tailscale.ipn.App
 import com.tailscale.ipn.ui.util.set
 import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.flow.StateFlow
 
-abstract class MDMSetting<T>(defaultValue: T, val key: String, val localizedTitle: String) {
+// Don't keep track of the key if shouldRegisterKey is false - shouldRegisterKey should be false when creating an MDMSetting for retrieving a setting.
```
Comment (agottardo) [inline android/src/main/java/com/tailscale/ipn/mdm/MDMSettingsDefinitions.kt:13]: Does this mean that all initializer calls in MDMSettings.kt need to be updated with `shouldRegisterKey: false`?. The default value here is `true`.

File: android/src/main/java/com/tailscale/ipn/mdm/MDMSettingsDefinitions.kt (hunk 1 above)
Comment (kari-ts) [reply android/src/main/java/com/tailscale/ipn/mdm/MDMSettingsDefinitions.kt:29]: we are using this in App.getSyspolicyStringValue - MDMSettings has all the MDMSettings we've created, and we want to check to see if there is one with the key we're given

File: android/src/main/java/com/tailscale/ipn/mdm/MDMSettingsDefinitions.kt (hunk 2 above)
Comment (kari-ts) [reply android/src/main/java/com/tailscale/ipn/mdm/MDMSettingsDefinitions.kt]: no, they should register keys. when we are checking to see if there exists an MDMSetting for a certain key, shouldRegisterKey should be false. I updated the comment to hopefully be a little clearer

File: libtailscale/syspolicy_handler.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+import (
+	"log"
+
+	"tailscale.com/util/syspolicy"
+)
+
+// androidHandler is a syspolicy handler for the Android version of the Tailscale client,
+// which lets the main networking code read values set via the Android RestrictionsManager.
+type androidHandler struct {
```
Comment (oxtoacart) [inline libtailscale/syspolicy_handler.go:14]: nit: since this isn't in its own package, a more specific name would be good, like `syspolicyHandler`.

**Review by agottardo — APPROVED** [review]

**Review by oxtoacart — COMMENTED** [review]
[This](https://github.com/tailscale/tailscale-android/pull/308) is a simpler way to do it.

**Review by oxtoacart — APPROVED** [review]

**Pull request merged**